examples/viewer.js

```diff
@@ -223,18 +473,127 @@ async function startViewer(localView, remoteView, formValues, onStatsReport, onR
             iceTransportPolicy: formValues.forceTURN ? 'relay' : 'all',
         };
         viewer.peerConnection = new RTCPeerConnection(configuration);
+
+        viewer.peerConnection.onicegatheringstatechange = (event) => {
+            if (viewer.peerConnection.iceGatheringState === 'gathering') {
+                metrics.viewer.iceGathering.startTime = Date.now();
+            } else if (viewer.peerConnection.iceGatheringState === 'complete') {
+                metrics.viewer.iceGathering.endTime = Date.now();
+            }
+        };
+
+        viewer.peerConnection.onconnectionstatechange = (event) => {
+            if (viewer.peerConnection.connectionState === 'new' || viewer.peerConnection.connectionState === 'connecting') {
+                metrics.viewer.peerConnection.startTime = Date.now();
+            }
+            if (viewer.peerConnection.connectionState === 'connected') {
+                metrics.viewer.peerConnection.endTime = Date.now();
+                metrics.viewer.ttffAfterPc.startTime = metrics.viewer.peerConnection.endTime;
+            }
+        };
+
+        viewer.peerConnection.oniceconnectionstatechange = (event) => {
+            if (viewer.peerConnection.iceConnectionState === 'connected') {
+                viewer.peerConnection.getStats().then(stats => {
+                    stats.forEach(report => {
+                        if (report.type === 'candidate-pair') {
+                            activeCandidatePair = report;
+                        }
+                    });
+                });
+            }
+        };
+
         if (formValues.openDataChannel) {
             const dataChannelObj = viewer.peerConnection.createDataChannel('kvsDataChannel');
             viewer.dataChannel = dataChannelObj;
-            dataChannelObj.onopen = event => {
-                dataChannelObj.send("Opened data channel by viewer");
+            dataChannelObj.onopen = () => {
+                if (formValues.enableDQPmetrics) {
+                    dataChannelLatencyCalcMessage.timestamp1 = Date.now();
+                    dataChannelObj.send(JSON.stringify(dataChannelLatencyCalcMessage));
+                } else {
+                    dataChannelObj.send("Opened data channel by viewer");
+                }
             };
             // Callback for the data channel created by viewer
-            dataChannelObj.onmessage = onRemoteDataMessage;
+            let onRemoteDataMessageViewer = (message) => {
```

Comment:
What happens if the C SDK does not enable this? As I understand, master end sending is disabled by default in C SDK. I think we should have a checkbox specific to this. It should be separate from the DQP checkbox. C SDK customers could customize the data channel application to not send these or even not enable it, but still want viewer side metrics.

Reply:
Introduced a new checkbox and decoupled this from DQP. Also tested locally in all modes (DQP + Datachannel benchmarking, only DQP, only Datachannel benchmarking, none) and everything works fine
<img width="484" alt="Screenshot 2023-12-19 at 9 06 27 PM" src="https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-js/assets/27426488/d57468ee-86fa-45ff-bd29-ca42bba6683a">